015-kafka-topic-encryption.md

```diff
@@ -0,0 +1,133 @@
+This document proposes a technical solution for providing upper-layer encryption-at-rest for Kafka systems.
+
+
+## Proposal
+
+An implementation of topic encryption is proposed whereby the message stream between client and broker is
+intercepted. Incoming data messages (Produce requests) are inspected to determine whether their payload
+should be encrypted according to a policy.  If so, the data portions are encrypted and the modified
+message is forwarded to the broker. As a result, the topic data is stored by the broker in encrypted form.
+On the reverse direction, encrypted responses (responses to Fetch requests) are decrypted prior to being sent to clients.
+
+As defined in the encryption policy, each topic can be encrypted by a different key,
+allowing brokers to store a mix of encrypted and unencrypted data, where
+data owners can manage the keys to their topics.
+Keys ideally are stored in a key management system with access policies and logging. 
+
+A core topic-encryption component, termed the _Encryption Module_, is proposed which is then deployed in a proxy. 
+
+The diagram below depicts the main components of the proposal, illustrating clients sending and receiving plaintext messages while the proxy exchanges ciphertext messages with the broker:
+
+![overview](images/015-kafkaenc-overview.png)
+
+### Encryption Module
+The Encryption Module is the top-level component which encapsulates encryption functionality and is embedded in a proxy.
+A central design goal is adaptability through modularity. Encrypter/decrypter(s), the policy metadata service and the key management service (KMS) are expressed as interfaces and pluggable. The Encryption Module instantiates these components during initialization, allowing a configurable combination of KMS, metadata and encryption implementations.
+
+#### Encrypter-Decrypter
+The Encrypter-Decrypter sub-component performs the actual encryption and decryption of messages, implementing a particular encryption algorithm and making use of the KMS service to obtain topic keys.
+
+#### Key Service
+The Key Service is used in the Encryption Module to handle interactions with a key management system and to perform optional primitives such as those related to envelope encryption. The interface is generic in order to accommodate various key management implementations.
+
+#### Policy Service
+Encrypter-decrypters require a means to lookup encryption requirements and parameters for a topic. As a corresponding standard does not exist, the core component will interact with a generic internal interface for retrieving policy on how a topic is, or not is, to be encrypted. Implementers can develop and configure their own implementation of the policy service, be it a remote microservice, a local pre-configured cache, a cluster registry, etc.
+
+For each topic to be encrypted, a policy exists detailing:
+
+- topic name
```

Comment:
I can see use cases where it would be convenient to have a single policy governing multiple topics, e.g. all topics beginning with a given prefix should use the same policy.

Reply:
I Kafka record is comprised of Key and Value fields. Encryption is applied to the Value field only. The message key therefore is the same across encryptions.  

As an example, the Kafka javadoc for ProducerRecord defines Key and Value as distinct fields:
https://kafka.apache.org/10/javadoc/org/apache/kafka/clients/producer/ProducerRecord.html

Reply:
We have considered supporting a regular expression for topic names. Because regular expressions can result in multiple matches and thus can be a bit unpredictable, we initially have taken a conservative stance that either policy is explicitly specified for each topic or a global wildcard (*) policy is applied.  We are definitely open to discussion. We have also considered the idea of a default policy.  Supporting multiple topics does seem to make sense either as a comma-separated list of topic names or the slightly risky regex.

Reply:
We leave open the full functionality of the policy service. The matching expression does not have to be an exact match or a regular expression but rather also some sort of query expression. Same caveat with regular expressions apply as mentioned above apply - a topic may match more than one expression and therefore more than one policy.  The convenience of regular and query expressions must be weighed with potential ambiguities which may arise.

Reply:
I agree that things like regexes or prefixes would need to be handled carefully since they open a big space for misconfiguration.